submissions/expero1/oop-exercise/index.js

```diff
@@ -0,0 +1,105 @@
+/* Refer to https://github.com/OleksiyRudenko/a-tiny-JS-world for the task details
+   Complete the below for code reviewers' convenience:
+
+   Code repository: _put repo URL here_
+   Web app: _put project's github pages URL here_
+   */
+
+// ======== OBJECTS DEFINITIONS ========
+// Define your objects here
+
+class Habitat {
+  static SPECIES = { DOG: "dog", CAT: "cat", HUMAN: "human" };
+  static GENDER = { MALE: "male", FEMALE: "female" };
+  friends = [];
+  PROPS = ["species", "name", "gender", "legs", "saying"];
+  constructor(species, name, gender, legs, saying) {
+    this.species = species;
+    this.name = name;
+    this.gender = gender;
+    this.legs = legs;
+    this.saying = saying;
+  }
+
+  addFriend(habitat) {
+    if (!this.friends.includes(habitat)) {
+      this.friends.push(habitat);
+    }
+    if (!habitat.friends.includes(this)) {
+      habitat.friends.push(this);
+    }
+  }
+
+  toString() {
+    let propsKeys = this.PROPS.map((key) => {
+      return this[key];
+    });
+    let info = propsKeys.join("; ");
+    if (this.friends.length > 0) {
+      info += "; ";
+      let friendList = this.friends.reduce((list, friend) => {
+        list.push(friend.species + " " + friend.name);
+        return list;
+      }, []);
+
+      info += friendList.join(", ");
+    }
```

Comment:
A use case for `Array#map`, as here we convert every single array element into correlating data. You won't need extra variable. `join` here would be perfectly chainable.

Also, if this conversion is done before `let info = propsKeys.join("; ");` and its result is made a part of `propKeys` array then `propsKeys.join("; ")` will join all array elements, including the list of friends.

Perhaps, you will be able to get rid of `info` variable as well.

Reply:
Thank you for your comment. You are right, after the refactoring, the code is much cleaner and easier to read.